armory/utils/export.py

```diff
@@ -92,369 +75,536 @@ def _make_output_dir(self):
-        self, x, x_adv, y=None, y_pred_adv=None, y_pred_clean=None, classes_to_skip=None
-    ):
-
-        plot_boxes = True if y is not None else False
-        if classes_to_skip is not None:
-            if isinstance(classes_to_skip, int):
-                classes_to_skip = [classes_to_skip]
 
-        for x_i, x_adv_i in zip(x, x_adv):
+class ImageClassificationExporter(SampleExporter):
+    def _export(self, x, x_adv=None, y=None, y_pred_adv=None, y_pred_clean=None):
+        for i, x_i in enumerate(x):
+            self._export_image(x_i, name="benign")
 
-            if self.saved_samples == self.num_samples:
-                break
+            # Export adversarial image x_adv_i if present
+            if x_adv is not None:
+                x_adv_i = x_adv[i]
+                self._export_image(x_adv_i, name="adversarial")
 
-            assert np.all(
-                x_i.shape == x_adv_i.shape
-            ), f"Benign and adversarial images are different shapes: {x_i.shape} vs. {x_adv_i.shape}"
-            if x_i.min() < 0.0 or x_i.max() > 1.0:
-                log.warning("Benign image out of expected range. Clipping to [0, 1].")
-            if x_adv_i.min() < 0.0 or x_adv_i.max() > 1.0:
-                log.warning(
-                    "Adversarial image out of expected range. Clipping to [0, 1]."
-                )
+            self.saved_samples += 1
+        self.saved_batches += 1
+
+    def _export_image(self, x_i, name="benign"):
+        self.image = self.get_sample(x_i)
+        self.image.save(
+            os.path.join(self.output_dir, f"{self.saved_samples}_{name}.png")
+        )
+        if x_i.shape[-1] == 6:
+            self.depth_image = self.get_sample(x_i[..., 3:])
+            self.depth_image.save(
+                os.path.join(self.output_dir, f"{self.saved_samples}_depth_{name}.png")
+            )
 
-            if x_i.shape[-1] == 1:
-                mode = "L"
-                x_i_mode = np.squeeze(x_i, axis=2)
-                x_adv_i_mode = np.squeeze(x_adv_i, axis=2)
-            elif x_i.shape[-1] == 3:
-                mode = "RGB"
-                x_i_mode = x_i
-                x_adv_i_mode = x_adv_i
-            elif x_i.shape[-1] == 6:
-                mode = "RGB"
-                x_i_mode = x_i[..., :3]
-                x_adv_i_mode = x_adv_i[..., :3]
-                x_i_depth = x_i[..., 3:]
-                depth_image = Image.fromarray(
-                    np.uint8(np.clip(x_i_depth, 0.0, 1.0) * 255.0), mode
-                )
-                depth_image.save(
-                    os.path.join(self.output_dir, f"{self.saved_samples}_depth.png")
-                )
-            else:
-                raise ValueError(f"Expected 1 or 3 channels, found {x_i.shape[-1]}")
+    @staticmethod
+    def get_sample(x_i):
+        if x_i.min() < 0.0 or x_i.max() > 1.0:
+            log.warning("Image out of expected range. Clipping to [0, 1].")
+
+        # Export benign image x_i
```

Comment:
Add some documentation here w.r.t. what shapes are expected and why. (E.g., grayscale, RGB, RGB-Depth) Also, that this expects a floating point [0.0, 1.0] input.